src/main/java/io/rcktapp/api/handler/sql/SqlDb.java

```diff
@@ -305,76 +310,123 @@ public void reflectDb() throws Exception
+                        }
+                    }, dbMetadataExecutorPool));
+
+                }
             }
-            colsRs.close();
 
-            ResultSet indexMd = dbmd.getIndexInfo(apiConn.getCatalog(), null, tableName, true, false);
-            while (indexMd.next())
-            {
-               String colName = indexMd.getString("COLUMN_NAME");
-               Column col = getColumn(tableName, colName);
-               col.setUnique(true);
+            log.info("{} building foreign key relationships", getType());
+            //-- now link all of the fks to pks
+            //-- this is done after the first loop
+            //-- so that all of the tbls/cols are
+            //-- created first and are there to
+            //-- be connected
+            List<CompletableFuture<?>> keyFutures = new ArrayList<>();
+            try (ResultSet foreignKeyTablesRS = dbmd.getTables(null, "public", "%", new String[]{"TABLE"}))  {
+                while (foreignKeyTablesRS.next())
+                {
+                    String tableName = foreignKeyTablesRS.getString("TABLE_NAME");
+
+                    if (Stream.of(ignoreTablePrefixes.split(",")).filter(Predicate.not(String::isEmpty)).anyMatch(tableName::startsWith)) continue;
+
+                    keyFutures.add(CompletableFuture.supplyAsync(() -> {
+                        try (Connection keyConnection = DriverManager.getConnection(getUrl(), getUser(), getPass())) {
+                            DatabaseMetaData databaseMetaData = keyConnection.getMetaData();
+
+                            try (ResultSet keyMd = databaseMetaData.getImportedKeys(keyConnection.getCatalog(), null, tableName)) {
+                                while (keyMd.next())
+                                {
+                                    String fkTableName = keyMd.getString("FKTABLE_NAME");
+                                    String fkColumnName = keyMd.getString("FKCOLUMN_NAME");
+                                    String pkTableName = keyMd.getString("PKTABLE_NAME");
+                                    String pkColumnName = keyMd.getString("PKCOLUMN_NAME");
+
+                                    Column fk = null;
+                                    Table fkTable = tableFutures.get(fkTableName).get();
+                                    for (Column c : fkTable.getColumns())
+                                        if (c.getName().equalsIgnoreCase(fkColumnName))
+                                            fk = c;
+                                    if (fk == null)
+                                        throw new RuntimeException("fk column not found: " + fkTableName + "." + fkColumnName);
+                                   Table pkTable = tableFutures.get(pkTableName).get();
+                                    for (Column pk : pkTable.getColumns())
+                                        if (pk.getName().equalsIgnoreCase(pkColumnName))
+                                            fk.setPk(pk);
```

Comment:
The previous uses of 'getColumn' required the 'tables' list to already be populated.  Any expectation of that in this asynchronous situation was definitely problematic.  Instead, I grab the table future and use that to scan for the column object they're looking for here.  The effective pattern is that all the getImportedKeys resultsets buffer up waiting for all the getTable resultsets to finish being retrieved, so once the tables resolve the remainder of the importedKeys loop interations finish very quickly.